### ISSUE TITLE
[production] ChangeNotifier.removeListener: Null check operator used on a null value

### ISSUE THREAD

**Issue by kinex** [flutter/flutter#71983]
I see a lot of these exceptions in Crashlytics for my app:
```bash
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: Null check operator used on a null value. Error thrown Instance of 'ErrorSummary'.
       at ChangeNotifier.removeListener(ChangeNotifier.java:179)
       at EditableTextState.dispose(EditableTextState.java:1590)
       at StatefulElement.unmount(StatefulElement.java:4854)
       at _InactiveElements._unmount(_InactiveElements.java:1998)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at Iterable.forEach(Iterable.java)
       at _RenderChipElement.visitChildren(_RenderChipElement.java:2079)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at MultiChildRenderObjectElement.visitChildren(MultiChildRenderObjectElement.java:6218)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at MultiChildRenderObjectElement.visitChildren(MultiChildRenderObjectElement.java:6218)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at MultiChildRenderObjectElement.visitChildren(MultiChildRenderObjectElement.java:6218)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at _SliverPersistentHeaderElement.visitChildren(_SliverPersistentHeaderElement.java:270)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at _InactiveElements.<fn>(_InactiveElements.java:1996)
       at ComponentElement.visitChildren(ComponentElement.java:4681)
       at _InactiveElements._unmount(_InactiveElements.java:1994)
       at ListIterable.forEach(ListIterable.java)
       at _InactiveElements._unmountAll(_InactiveElements.java:2007)
       at BuildOwner.<fn>(BuildOwner.java:2821)
       at BaseMouseTracker._deviceUpdatePhase(BaseMouseTracker.java:317)
       at BuildOwner.finalizeTree(BuildOwner.java:2820)
       at WidgetsBinding.drawFrame(WidgetsBinding.java:915)
       at RendererBinding._handlePersistentFrameCallback(RendererBinding.java:302)
       at SchedulerBinding._invokeFrameCallback(SchedulerBinding.java:1117)
       at SchedulerBinding.handleDrawFrame(SchedulerBinding.java:1055)
       at SchedulerBinding._handleDrawFrame(SchedulerBinding.java:971)
```
Stack trace is from Android, but there is same issue in iOS. I have no idea where in my code this exception happens.
```bash
[√] Flutter (Channel stable, 1.22.4, on Microsoft Windows [Version 10.0.19041.685], locale en-US)
    • Flutter version 1.22.4 at C:\tools\fluttersdk\flutter
    • Framework revision 1aafb3a8b9 (4 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at c:\tools\android-sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = C:\tools\android-sdk
    • ANDROID_SDK_ROOT = C:\tools\android-sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.51.1)
    • VS Code at C:\Users\me\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.17.0

[√] Connected device (1 available)
    • SM G930F (mobile) • ad0c16039914a951e1 • android-arm64 • Android 8.0.0 (API 26)
```

**Comment by TahaTesser** [comment]
Hi @kinex 
Can you please provide your `flutter doctor -v` **with verbose**, your `crash logs`, your `pubspec.yaml`, the list of affected devices, and if possible your `flutter run --release --verbose` running on the same OS version as the affected devices.
Thank you

**Comment by kinex** [comment]
I updated `flutter doctor -v` output. I think `pubspec.yaml` or the list of affected devices are not important (big app, many pubspec.yaml files and this happens both in Android and iOS). Most probably this is of course an issue in my own code, but I haven't found it yet. This may be related code on my side (I cannot be sure though, as the stack trace does not include my code):

```dart
typedef double OnSaveValue(double value);
typedef String OnValidate(double value);

/// Generic widget for editing a numeric field.
class NumericEditField extends StatefulWidget {
  const NumericEditField(
      {Key key,
      this.controller,
      this.initializeControllerText = false,
      this.focusNode,
      this.nextFocusNode,
      this.textInputAction,
      this.labelText = '',
      this.suffixText,
      this.notValidMessage = '',
      this.decimalDigits = 2,
      this.formatAsDecimal = true,
      this.maxLength = 16,
      this.signed = true,
      this.initialValue,
      this.onSaveValue,
      this.saveValueOnFocusLost = true,
      this.saveValueOnSubmit = false,
      this.validator,
      this.isReadOnly = false,
      this.textAlign = TextAlign.start,
      this.icon,
      this.suffixIcon,
      this.showSuffixIconOnlyIfEmpty = false,
      this.helperText})
      : super(key: key);

  /// Parent may want to pass [controller] if it needs to update text outside
  /// this widget. Caller must set initial value for controller.text or set
  /// [initializeControllerText] to true.
  final TextEditingController controller;

  /// True to initialize text of passed [controller]. Ignored if [controller]
  /// is null.
  final bool initializeControllerText;

  final FocusNode focusNode;
  final FocusNode nextFocusNode;
  final TextInputAction textInputAction;

  final String labelText;
  final String suffixText;
  final String notValidMessage;
  final int decimalDigits;
  final bool formatAsDecimal;
  final int maxLength;
  final bool signed;

  /// Initial value. Ignored if custom [controller] has been passed
  /// (caller is responsible for initializing [controller] in that case).
  final double initialValue;

  // Callback to call when a value is saved. Value returned by callback
  // is set to the text of the field (if not null).
  final OnSaveValue onSaveValue;

  final bool saveValueOnSubmit;
  final bool saveValueOnFocusLost;
  // Function to validate input. Should return error message string or null
  // if validation passed.
  final OnValidate validator;
  final bool isReadOnly;
  final TextAlign textAlign;
  final Widget icon;

  // suffix icon
  final Widget suffixIcon;

  // whether to show suffix icon only if edit field is empty
  final bool showSuffixIconOnlyIfEmpty;

  // helper text
  final String helperText;

  @override
  _NumericEditFieldState createState() {
    return _NumericEditFieldState();
  }
}

class _NumericEditFieldState extends State<NumericEditField> {
  TextEditingController _controller;
  FocusNode _focusNode;

  @override
  void initState() {
    super.initState();
    _controller = widget.controller ?? TextEditingController();
    if (widget.controller == null || widget.initializeControllerText) {
      _controller.text = StringUtil.toNumericEditFieldText(widget.initialValue,
          decimalDigits: widget.decimalDigits,
          emptyIfZero: true,
          formatAsDecimal: widget.formatAsDecimal);
    }
    _focusNode = widget.focusNode ?? FocusNode();
    _focusNode.addListener(_onFocusChanged);
  }

  @override
  void dispose() {
    _focusNode.removeListener(_onFocusChanged);
    if (widget.focusNode == null) {
      // dispose only if created locally
      _focusNode.dispose();
    }
    _focusNode = null;
    if (widget.controller == null) {
      _controller.dispose();
    }
    _controller = null;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final invalidDecimalSeparator =
        NumberFormatter.decimalSeparator == ',' ? '.' : ',';
    return TextFormField(
      controller: _controller,
      focusNode: _focusNode,
      textInputAction: widget.textInputAction,
      decoration: InputDecoration(
          icon: widget.icon,
          labelText: widget.labelText,
          hintText: NumberFormatter.toDecimal(0, widget.decimalDigits),
          helperText: widget.helperText,
          suffixText: widget.suffixText,
          suffixIcon: widget.showSuffixIconOnlyIfEmpty &&
                  (_focusNode?.hasFocus == true ||
                      _controller?.text?.isNotEmpty == true)
              ? null
              : widget.suffixIcon,
          errorMaxLines: 16,
          border: widget.isReadOnly
              ? InputBorder.none
              : Theme.of(context).inputDecorationTheme.border),
      inputFormatters: [
        LengthLimitingTextInputFormatter(widget.maxLength),
        widget.decimalDigits == 0
            ? FilteringTextInputFormatter.digitsOnly
            : FilteringTextInputFormatter.deny(invalidDecimalSeparator,
                replacementString: NumberFormatter.decimalSeparator)
      ],
      textAlign: widget.textAlign,
      enabled: !widget.isReadOnly,
      // ignore "signed" in iOS as it shows full keyboard
      keyboardType: TextInputType.numberWithOptions(
          signed: widget.signed && !isIOS, decimal: widget.decimalDigits > 0),
      validator: (value) {
        if (value == '') {
          value = '0';
        }
        final d = StringUtil.parseDouble(value, widget.formatAsDecimal);
        if (d == null) {
          return widget.notValidMessage;
        }
        if (widget.validator != null) {
          return widget.validator(d);
        }
        return null;
      },
      onFieldSubmitted: (value) {
        if (widget.saveValueOnSubmit) {
          _saveValue(value);
        }
        if (widget.textInputAction == TextInputAction.next) {
          // nextFocus does not work in all cases, so allow to specify
          // next FocusNode if needed
          if (widget.nextFocusNode != null) {
            widget.nextFocusNode.requestFocus();
          } else {
            FocusScope.of(context).nextFocus();
          }
        } else if (widget.textInputAction == TextInputAction.done) {
          FocusScope.of(context).unfocus();
        }
      },
      onSaved: (value) {
        _saveValue(value);
      },
    );
  }

  void _onFocusChanged() {
    if (_focusNode == null || _controller == null) {
      return;
    }
    if (_focusNode.hasFocus) {
      // TODO: is this still needed https://github.com/flutter/flutter/issues/11416
      _moveCaretToEnd();
    } else {
      // User may navigate to other field without pressing the "Done" button,
      // so update field when focus lost to update dependent fields
      if (widget.saveValueOnFocusLost) {
        _saveValue(_controller.text);
      }
    }
  }

  void _moveCaretToEnd() {
    _controller.selection =
        TextSelection.collapsed(offset: _controller.text.length);
  }

  void _saveValue(String value) {
    if (_controller == null) {
      return;
    }

    final d = StringUtil.parseDouble(value, widget.formatAsDecimal) ?? 0;

    final newValue = widget.onSaveValue != null ? widget.onSaveValue(d) : d;
    if (newValue != null) {
      // format value correctly
      // callback handler may have changed the value
      _controller.text = StringUtil.toNumericEditFieldText(newValue,
          decimalDigits: widget.decimalDigits,
          emptyIfZero: true,
          formatAsDecimal: widget.formatAsDecimal);
    }
  }
}

```

**Comment by TahaTesser** [comment]
Hi @kinex 
Can you please provide complete `flutter run --verbose` logs, and a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example), not production code?
Thank you

**Comment by kinex** [comment]
I cannot reproduce this issue. It happens on users' devices.

**Comment by TahaTesser** [comment]
Hi @kinex 
Can you please provide your `crash logs`, your `pubspec.yaml`, the list of affected devices and if possible your `flutter run --release --verbose` running on the same OS version as the affected devices.
Thank you

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.

**Comment by FrantisekGazo** [comment]
@TahaTesser @pedromassangocode I encounter this error as well. It's caused by using the ChangeNotifier after dispose() call. 

The issue is that you set `_listeners = null` in `dispose()` and use `_listeners!` in `hasListeners`/`addListener()`/... => which will throw this error if null is set.

The best would be to do nothing if dispose() was called instead of throwing errors. **But if that's not possible can you at least add `bool get isDisposed` getter so that it can be checked?**

**Comment by TahaTesser** [comment]
@FrantisekGazo 
Thanks for the details and confirmation, very helpful.  Are you on the stable channel as well? Can you please provide your `flutter doctor -v`?

@kinex
Change notifier could be part of third-party packages as well, which might not have properly handled it but now we know confirmation and explanation. 

Thank you

**Comment by FrantisekGazo** [comment]
```
[✓] Flutter (Channel stable, 1.22.5, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.22.5 at /Users/franky/CODE/flutter-sdk
    • Framework revision 7891006299 (6 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/franky/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/franky/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.8.4

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio 4.0.app/Contents
    • Flutter plugin installed
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86 • Android 7.0 (API 24) (emulator)
    ! Device emulator-5556 is offline.

! Doctor found issues in 1 category.
```

**Comment by kinex** [comment]
I think it is better to close this until I verify if the issue still exists. I found some very old and stupid code in my app. It was modifying the text of a `TextField` in the `build` method of a widget which is not allowed. So this may cause the crash. To verify the fix I will publish an updated app soon, I will then see from Crashlytics if the issue still exists.

**Comment by FrantisekGazo** [comment]
please reopen this issue @TahaTesser . 

The `ChangeNotifier` in the latest flutter version is still the same, so the issue with calling `_listeners!` on null value is still there (my original comment [here](https://github.com/flutter/flutter/issues/71983#issuecomment-762108820))

**Comment by kinex** [comment]
Yes it looks this problem still occurs in my app too even after fixing the code I described above. I have also upgraded to Flutter 2.0.

**Comment by aboutlan** [comment]
I have the same problem.

**Comment by slavap** [comment]
@TahaTesser 
Please fix it as proposed above "**at least add bool get isDisposed getter so that it can be checked**"
or better remove all _listeners! from the code, so it never crashes In many cases it is absolutely impossible to control if ChangeNotifier was already disposed, e.g. I'm having problems with disposed ScrollPosition by third party code.

**Comment by caseycrogers** [comment]
Yeah the current implementation is really bad.

You should always be cleaning up your listeners, so you should put `removeListener` in your `dispose` function. But, you can't do that in a lot of cases where you don't know ahead of time if the change notifier or the stateful widget will be disposed first. The only option in these cases is to wrap your call to `removeListener` in a `try...on` block which is really poor practice.

**Comment by caseycrogers** [comment]
Okay, so it took an ENORMOUS amount of work, but I've created "minimal" reproduction code for the specific way of triggering this bug that was appearing in my app. Sorry for the length and complexity of the reproduction, this is a pretty subtle and messy bug. See [gist](https://gist.github.com/caseycrogers/c018ac774645be6619f9ac4de3c6d2dd).

Here's the repro steps (what my gist is doing to trigger the bug):
1. Create a `StatefulWidget` that listens to a `ScrollPositionWithSingleContext` accessed via `Scrollable.of(context)` and calls `removeListener` in it's dispose function. We'll call this widget `ScrollAware`.
2. Create a widget that toggles between two child apps. eg a stateful widget that builds child A if a bool is true, child B otherwise. We'll call this widget `Toggler`.
3. Write an app that produces the following widget state:

```
Toggler > Navigator > [
  Page > ListView > [
    ScrollAware,
    Container(height: 1000),
  ], 
]```

4. Push an arbitrary page to the navigator stack so that our `ListView` is no longer in an active route:

```
Toggler > Navigator > [
  Page >ListView > [
    ScrollAware,
    Container(height: 1000),
  ], 
  Page > Container,
]```

5. Toggle the toggler such that the navigator is replaced by an arbitrary widget and is thus no longer in the widget tree:

```
Toggler > Container```

6. Note that the `removeListener` exception is thrown.

What's happening is that the position change notifier is disposed when the enclosing route is no longer current.

This is particularly pernicious because the order in which `dispose` is called depends on user behavior. In the above example, the position notifier was disposed of first. But, you scroll down the list view, the scroll aware widget will be disposed as it leaves the screen and thus will be disposed before the position change notifier.

I'm fairly certain this entire mess is a result of a bug in `ListView/Scrollable` (it shouldn't dispose of the position until the widget is disposed), so maybe this isn't the most ironclad example. However, I think this is still a decent example of why it's bad that calls to `removeListener` are unsafe if the `ChangeNotifier` is being provided by an external code source.

What if the above issue weren't a result of a bug, and were a result of intended behavior? Here I can use `ModalRoute.of(context)!.isCurrent` to intuit whether or not the call to `removeListener` will be safe, but that's pretty gymnastic and might not work in different circumstances.

It's seems clear that there's a strong motive on the `flutter/flutter` team for not making calls to `removeListener` safe and/or exposing whether or not the change notifier has been disposed. Can that motive be shared and defended here, or better yet, in the source code's documentation?

CC @goderbauer here is an example where the user doesn't know if calling `removeListener` is safe.

**Comment by caseycrogers** [comment]
In the meantime, here's a solution I recommend for others facing similar issues with `removeListener` (only ~75% joking):

```dart
extension IsDisposedChangeNotifier on ChangeNotifier {
  bool get isDisposed {
    try {
      hasListeners;
    } on AssertionError {
      return true;
    }
    return false;
  }
}
```

**Comment by slightfoot** [comment]
@TahaTesser I am looking into this issue as part of a PR. And I realise as of Flutter v3.0.0.. its fixed.
It was fixed in Feb by [#97988 ](https://github.com/flutter/flutter/pull/97988) .

Closing.

**Comment by TahaTesser** [comment]
> @TahaTesser I am looking into this issue as part of a PR. And I realise as of Flutter v3.0.0.. its fixed.
> It was fixed in Feb by https://github.com/flutter/flutter/pull/97988.
> 
> Closing.

Thanks for the update!
